gpytorch/priors/prior.py

```diff
@@ -0,0 +1,66 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+from __future__ import unicode_literals
+
+from abc import abstractmethod, abstractproperty
+import torch
+from torch.nn import Module
+
+
+class Prior(Module):
+    """
+    Base class for Priors in GPyTorch.
+    In GPyTorch, a parameter can be assigned a prior by passing it as the `prior` argument to
+    :func:`~gpytorch.module.register_parameter`. GPyTorch performs internal bookkeeping of priors,
+    and for each parameter with a registered prior includes the log probability of the parameter under its
+    respective prior in computing the Marginal Log-Likelihood (see e.g. :func:`~gpytorch.priors.Prior.forward`).
+
+    In order to define a new Prior in GPyTorch, a user must define at a minimum the following methods
+    * :func:`~gpytorch.priors.Prior.shape`, which returns the shape of the domain as a torch.Size.
+    * :func:`~gpytorch.priors.Prior.is_in_support`, which for a given parameter value returns a bool indicating
+        whether the parameter is contained in the support of the distribution.
+    * :func:`~gpytorch.priors.Prior._log_prob`, which returns the log-probability for a given parameter value
+        as a torch scalar.
+
+    """
+
+    @abstractmethod
+    def is_in_support(self, parameter):
+        raise NotImplementedError()
+
+    @property
+    def log_transform(self):
+        return self._log_transform
+
+    @abstractmethod
+    def _log_prob(self, parameter):
+        raise NotImplementedError()
+
+    @abstractproperty
+    def shape(self):
```

Comment:
Could we use "size" as the method name, to be consistent with other method names?

Reply:
I changed it from size to shape, but ok :)

Reply:
I think we should maintain shape. PyTorch implements a size *method* and a shape *property* that function the same way which I think we should be consistent with.

Reply:
Sounds good to me - I won't touch the LazyVariable part for now, opened #182 for this.